codegen-server/src/test/kotlin/software/amazon/smithy/rust/codegen/server/smithy/ConstraintsTest.kt

```diff
@@ -119,4 +125,9 @@ class ConstraintsTest {
         mapB.canReachConstrainedShape(model, symbolProvider) shouldBe true
         recursiveShape.canReachConstrainedShape(model, symbolProvider) shouldBe true
     }
+
+    @Test
+    fun `it should not consider shapes with the default trait as constrained`() {
+        structWithInnerDefault.canReachConstrainedShape(model, symbolProvider) shouldBe false
```

Comment:
I would call `isDirectlyConstrained` on `StructWithInnerDefault$PrimitiveBoolean` too, to make the unit tests be more unit-like.